Add tabs->favicon-scale setting

haasn opened the pull request:
This allows users to change the size of the favicon independently from
the size of the font/tab, in order to adjust the balance between
favicons and text. The drawing code is also adjusted to re-center the
favicons vertically with the text. The only other change is that the
initialization function also explicitly calls set_font() in order to
force the font/favicon size to be recomputed when the tab settings are
changed.

Works as expected even for values of 0.0 (which is equivalent to hiding
the favicon completely).

Closes #2549.

haasn commented:
Ah, seems like this manually needs to be added to `doc/help/settings.asciidoc` as well? I thought the documentation would be autogenerated

The-Compiler commented:
It is, but only if you run `tox -e docs` (or `python scripts/dev/src2asciidoc.py`) :wink: (I usually do that while merging if the PR author doesn't)

You'll also need to adjust the stubbed config for `test_small_icon_doesnt_crash` in `tests/unit/mainwindow/test_tabwidget.py` as it [fails currently](https://travis-ci.org/qutebrowser/qutebrowser/jobs/225829436#L732).

The-Compiler commented on qutebrowser/mainwindow/tabwidget.py:
```diff
@@ -79,6 +79,7 @@ def init_config(self):
         self.setTabPosition(position)
         tabbar.vertical = position in [QTabWidget.West, QTabWidget.East]
         tabbar.setSelectionBehaviorOnRemove(selection_behavior)
+        tabbar.set_font()
```
What's the intention of this? This `init_config` block only gets run for changes in the `tabs` section because of the `@config.change_filter('tabs')`. Changing the font is already handled in `TabBar` (`        config_obj.changed.connect(self.set_font)`)

haasn replied:
The intention is to get changes to `tabs->favicon-scale` to recompute the favicon size, which is done based on the font size. I guess I'll just move that out to its own function, though.

haasn commented:
Updated PR with changes:

- Moved the favicon-scale next to show-favicon, rather than the bottom
- Removed the `tabbar.set_font()` call from TabWidget. Instead, connected the favicon-size-computing code directly to `tabs->favicon-scale`
- Regenerated documentation

haasn commented:
No idea what the codecov bot is rambling about (my commit doesn't touch `qtnetworkdownloads.py`!), but the test suite passes now.

The-Compiler commented:
Don't worry about that - I think some tests in there sometimes take a different path because of timing issues - I should really check out what the issue is exactly and mark it as no coverage or something...

Something seems off here, but I don't know what - here's how things look with the default font settings and a 0.7 scale for me; the icon definitely isn't centered: 
![favicon](https://cloud.githubusercontent.com/assets/625793/25455033/bef66300-2ace-11e7-82e1-75eb737081c6.png)

I'm fine with merging it as-is if you can't figure it out either (as it's only something with small sizes - some rounding going wrong?), but maybe you have an idea.

haasn commented:
Seems like an off-by-1-pixel. Even at 1.0 scale it's consistently 1 pixel higher than before. Adding in an extra `+ 1` operand makes it match up with the current position at scale `1.0`.

Probably has something or other to do with the fact that out of the box the icon is already placed on `text_rect.top() + 1`, so recentering it to the text_rect without the `+ 1` operand causes it to be one pixel higher than before.

I've updated the commit to include the extra operand.

The-Compiler commented:
Seems to be broken :wink:

```
Traceback (most recent call last):
  File "/home/florian/proj/qutebrowser/git/qutebrowser/mainwindow/tabwidget.py", line 689, in subElementRect
    layouts = self._tab_layout(opt)
  File "/home/florian/proj/qutebrowser/git/qutebrowser/mainwindow/tabwidget.py", line 743, in _tab_layout
    icon_rect = self._get_icon_rect(opt, text_rect)
  File "/home/florian/proj/qutebrowser/git/qutebrowser/mainwindow/tabwidget.py", line 781, in _get_icon_rect
    icon_top = text_rect.center().y() + 1 - icon_rect.height() / 2
UnboundLocalError: local variable 'icon_rect' referenced before assignment
```

haasn commented:
Oops, of course. Have to use `tab_icon_size` before `icon_rect` is defined. Might as well also use the other constructor for QRect while I'm at it.

The-Compiler commented:
Thank you!
